### ISSUE TITLE
detect too-large blocks before inserting in DBS and reduce them

### ISSUE THREAD

**Issue by belforte** [dmwm/CRABServer#6670]
see #6656 

Modifty https://github.com/dmwm/CRABServer/blob/master/src/python/Publisher/TaskPublish.py to:

- [x] record for every publication block size, time, status and size of intitial file list (which has the lumi info and is the largest component of the block)
- [x] see if I can figure out a threshold for size
- [x] compute file-list size in TaskPublish before inserting into DBS and if too large lower max_file_per_block from 100 to whatever sensible

**Comment by belforte** [comment]
it is a good time to also take care of #6616 
Goal is to get rid of "fake" Publication error reports in grafana to avoid wasting time in investigating.

**Comment by belforte** [comment]
recording code added in my branch https://github.com/belforte/CRABServer/tree/record-publication-size-and-time-for-6670
now under test in preprod via `./start.sh -g`

**Comment by belforte** [comment]
on Hold, waiting for more Publication errors to appear and check diagnostic information for them in the new file `/data/srv/Publisher/logs/STATS.txt`

**Comment by belforte** [comment]
here are publication lasting more than 100 sec in last days, unsurprisingly same user/task(s) as last errors, but initial side of request (filestructSize)does not look very telling, sigh
```
[crab3@crab-prod-tw02 logs]$ cat STATS.txt |egrep 'time=[0-9]' > LSTATS.txt
[crab3@crab-prod-tw02 logs]$ cat LSTATS.txt 
PUBSTAT: Nfiles= 410, filestructSize=  17MB, iter= 1, blockSize=   4MB, time=106s, status=OK, task=210624_184927:ktaeun_crab_MC_generation_B0ToKD_ToKPiPi_MuFilter_TuneCP5_13TeV-pythia8-evtgen_24Jun2021
PUBSTAT: Nfiles= 330, filestructSize=  14MB, iter= 1, blockSize=   4MB, time=112s, status=OK, task=210624_184927:ktaeun_crab_MC_generation_B0ToKD_ToKPiPi_MuFilter_TuneCP5_13TeV-pythia8-evtgen_24Jun2021
PUBSTAT: Nfiles= 330, filestructSize=  14MB, iter= 2, blockSize=   4MB, time=107s, status=OK, task=210624_184927:ktaeun_crab_MC_generation_B0ToKD_ToKPiPi_MuFilter_TuneCP5_13TeV-pythia8-evtgen_24Jun2021
PUBSTAT: Nfiles= 305, filestructSize=  13MB, iter= 1, blockSize=   4MB, time=112s, status=OK, task=210624_184927:ktaeun_crab_MC_generation_B0ToKD_ToKPiPi_MuFilter_TuneCP5_13TeV-pythia8-evtgen_24Jun2021
PUBSTAT: Nfiles= 305, filestructSize=  13MB, iter= 2, blockSize=   4MB, time=107s, status=OK, task=210624_184927:ktaeun_crab_MC_generation_B0ToKD_ToKPiPi_MuFilter_TuneCP5_13TeV-pythia8-evtgen_24Jun2021
PUBSTAT: Nfiles= 305, filestructSize=  13MB, iter= 3, blockSize=   4MB, time=148s, status=OK, task=210624_184927:ktaeun_crab_MC_generation_B0ToKD_ToKPiPi_MuFilter_TuneCP5_13TeV-pythia8-evtgen_24Jun2021
PUBSTAT: Nfiles= 207, filestructSize=   8MB, iter= 1, blockSize=   4MB, time=110s, status=OK, task=210624_184927:ktaeun_crab_MC_generation_B0ToKD_ToKPiPi_MuFilter_TuneCP5_13TeV-pythia8-evtgen_24Jun2021
PUBSTAT: Nfiles= 207, filestructSize=   8MB, iter= 2, blockSize=   4MB, time=102s, status=OK, task=210624_184927:ktaeun_crab_MC_generation_B0ToKD_ToKPiPi_MuFilter_TuneCP5_13TeV-pythia8-evtgen_24Jun2021
PUBSTAT: Nfiles= 408, filestructSize=  17MB, iter= 3, blockSize=   4MB, time=102s, status=OK, task=210624_184927:ktaeun_crab_MC_generation_B0ToKD_ToKPiPi_MuFilter_TuneCP5_13TeV-pythia8-evtgen_24Jun2021
PUBSTAT: Nfiles= 690, filestructSize=  29MB, iter= 3, blockSize=   4MB, time=104s, status=OK, task=210624_184927:ktaeun_crab_MC_generation_B0ToKD_ToKPiPi_MuFilter_TuneCP5_13TeV-pythia8-evtgen_24Jun2021
PUBSTAT: Nfiles= 112, filestructSize=   4MB, iter= 1, blockSize=   4MB, time=113s, status=OK, task=210624_184927:ktaeun_crab_MC_generation_B0ToKD_ToKPiPi_MuFilter_TuneCP5_13TeV-pythia8-evtgen_24Jun2021
PUBSTAT: Nfiles= 315, filestructSize=  13MB, iter= 1, blockSize=   4MB, time=148s, status=OK, task=210624_184927:ktaeun_crab_MC_generation_B0ToKD_ToKPiPi_MuFilter_TuneCP5_13TeV-pythia8-evtgen_24Jun2021
PUBSTAT: Nfiles= 315, filestructSize=  13MB, iter= 2, blockSize=   4MB, time=147s, status=OK, task=210624_184927:ktaeun_crab_MC_generation_B0ToKD_ToKPiPi_MuFilter_TuneCP5_13TeV-pythia8-evtgen_24Jun2021
PUBSTAT: Nfiles= 315, filestructSize=  13MB, iter= 3, blockSize=   4MB, time=111s, status=OK, task=210624_184927:ktaeun_crab_MC_generation_B0ToKD_ToKPiPi_MuFilter_TuneCP5_13TeV-pythia8-evtgen_24Jun2021
PUBSTAT: Nfiles= 173, filestructSize=   7MB, iter= 1, blockSize=   4MB, time=123s, status=OK, task=210624_184927:ktaeun_crab_MC_generation_B0ToKD_ToKPiPi_MuFilter_TuneCP5_13TeV-pythia8-evtgen_24Jun2021
PUBSTAT: Nfiles= 134, filestructSize=   5MB, iter= 1, blockSize=   4MB, time=131s, status=OK, task=210624_184927:ktaeun_crab_MC_generation_B0ToKD_ToKPiPi_MuFilter_TuneCP5_13TeV-pythia8-evtgen_24Jun2021
PUBSTAT: Nfiles=  75, filestructSize=   3MB, iter= 1, blockSize=   3MB, time=102s, status=OK, task=210624_184927:ktaeun_crab_MC_generation_B0ToKD_ToKPiPi_MuFilter_TuneCP5_13TeV-pythia8-evtgen_24Jun2021
[crab3@crab-prod-tw02 logs]$ 

```

**Comment by belforte** [comment]
looking better, `filestructSize` appears a very bad metric, for 1MB intitial structure the publication time ranges from immediate to 35 seconds
```
[crab3@crab-prod-tw02 logs]$ cat STATS.txt |grep ' 1MB'|sort -nr -k11|tail -1
PUBSTAT: Nfiles=1008, filestructSize=   1MB, iter=10, blockSize= 170KB, time=  0s, status=OK, task=210628_090831:jaebeom_crab_ChiBToUps1SG_8p16TeV_ptHat4_TuneCUETP8M1_pPbEmd_GENSIM_v1
[crab3@crab-prod-tw02 logs]$ cat STATS.txt |grep ' 1MB'|sort -n -k11|tail -1
PUBSTAT: Nfiles=  46, filestructSize=   1MB, iter= 1, blockSize=   2MB, time= 35s, status=OK, task=210624_184927:ktaeun_crab_MC_generation_B0ToKD_ToKPiPi_MuFilter_TuneCP5_13TeV-pythia8-evtgen_24Jun2021
[crab3@crab-prod-tw02 logs]$ 
```

**Comment by belforte** [comment]
I need to find a better indicator.
So am removing the OnHold flag. and wiill try to find out what makes e.g. blockSize to be much bigger than initial size. Maybe simply look at number of lumis ? It is possible that some blowing up just comes from different ways of listing them on FJR vs. DBS blockdump format.

**Comment by belforte** [comment]
added #lumis to stats via https://github.com/belforte/CRABServer/commit/957a3b7a883069cb505e2b52083f508abc44cad7 but lumi list in the initial structure (i.e. `dbsFiles` list) is already in same format as it will be in `blockDump` so I still do not see how block can get bigger. Old input json files are not saved, so can't reproduce cases like this
```
PUBSTAT: Nfiles=  46, filestructSize=   1MB, iter= 1, blockSize=   2MB, time= 35s, status=OK, task=210624_184927:ktaeun_crab_MC_generation_B0ToKD_ToKPiPi_MuFilter_TuneCP5_13TeV-pythia8-evtgen_24Jun2021
```
maybe simply a rounding effect ?
Let's see what number of lumis can tell

**Comment by belforte** [comment]
since new code is in, these are the case where publication time was >10 seconds
```
 Nfiles=  24, filestructSize=   1MB, lumis=  24000, iter= 1, blockSize=   1MB, time= 18s, status=OK, task=210624_184714
 Nfiles=  30, filestructSize=   1MB, lumis=  30000, iter= 1, blockSize=   1MB, time= 16s, status=OK, task=210624_184714
 Nfiles=  37, filestructSize=   1MB, lumis=  37000, iter= 1, blockSize=   1MB, time= 19s, status=OK, task=210624_184714
 Nfiles=   9, filestructSize= 398KB, lumis=   9000, iter= 1, blockSize= 403KB, time= 13s, status=OK, task=210624_184927
 Nfiles=   8, filestructSize= 354KB, lumis=   8000, iter= 1, blockSize= 358KB, time= 10s, status=OK, task=210624_184927
```
at first sight a cut at 10K lumis per block looks appealing

**Comment by belforte** [comment]
OTOH there are lines like this
```
Nfiles= 671, filestructSize=  28MB, lumis= 671000, iter= 3, blockSize=   4MB, time=  7s, status=OK, task=210629_182953:kmotaama_crab_DR_UpsilonToMuMuDstarToD0picrt_29-06-2021
```
where 670K lumis were digested w/o a problem. Somehow there's something special and harmful in publication from tasks like `210624_184927:ktaeun_crab_MC_generation_B0ToKD_ToKPiPi_MuFilter_TuneCP5_13TeV-pythia8-evtgen_24Jun2021` which I still do not pin point.

**Comment by belforte** [comment]
I keept finding indication that time to insert a block (and hence chance of failure) is not predictable
based on block size and number of lumis alone. Surely the presence of parents matters, but possibly also exactly what those parents look like.

I further suspect that when http calls timeout at 5min, it is possible that the full block was delivered to DBS server anyhow, and IIRC what @yuyiguo told me, server will keep working on the request and eventually insert all files in the block, regardless of the fact that cmsweb FE closed the connection. So I need to "wait a while" before checking if files are inserted in DBS and declare the http error a "false positive".
That means that I can't simply retry right after a faillure, but need to wait e.g. a couple hours.

In a way the best solution is back to "mark the files as to retry and count retry numbers". This requires some heavish changes to current code, I think a change to CRAB Server API is needed to properly handle retries.
And of course when the presente of retried files is detected the publish script can lower the number of files per block e.g. from 100 to 20, 5, 1 and after 3 attemps call it final. But is the most solid solution.

**Comment by belforte** [comment]
problem suddenly escalated during weekend.
There are a few odd tasks with tens of thousands of files which constantly timeout when writing to DBS/phys03
See story in:
https://cms-logbook.cern.ch/elog/GlideInWMS/7766

Put temporary kludge in production TW hacking code in the container.
Will also push to current production branch based on https://github.com/dmwm/CRABServer/tree/py2only
I have done the following
* introduce a list of users to skip. Allows to quickly take care of similar cases by restarting Publisher skipping bad user tasks so that good things go on while bad ones are looked at
* since all and only the bad tasks in this case have name like '2018UL' or 'UL2018' and I have seen same tasks from same users to give publication failures already, I force publication blocks to 10 files for those. With the they are still slow, but at least complete.

**Comment by belforte** [comment]
I have made PR's with the above hot fix to both py2only and master branches https://github.com/dmwm/CRABServer/pull/6745 https://github.com/dmwm/CRABServer/pull/6746
And added list of users to skip to PublisherConfig.py template in puppet: https://gitlab.cern.ch/ai/it-puppet-hostgroup-vocmsglidein/-/commit/0b437ab8e5c9c461f5f9a27d3f4c05ce44d72d68

**Comment by belforte** [comment]
the above PR is now in tag py2only.2 and deployed in preprod

**Comment by belforte** [comment]
new idea: computer #lumis/#files and use it to get #lumis per block and limit files-per-block to keep lumis per block below 1M (or less)
Looking at past stat, this cut selects of cases wheren publication failes after >100sec, even if there a are a handful "false positive" which would have been splitted in smaller blocks w/o need... no big damage.
Below, from `logs/STATS.txt` the last 100 lines of a CSV file showing for each publication attempt
#files, #lumis, time, status, #lumis/#files and sorted by #lumis/#files.
COmmands:
```
cat STATS.txt |awk -F'[,=]' '{print $2","$6","$12","$14","$6/$2}' > FLTS.csv
cat FLTS.csv |sort -n -t, -k5|tail -100
```
output
```
[crab3@crab-prod-tw01 tmp]$ cat FLTS.csv |sort -n -t, -k5|tail -100
   3,  60000,  6s,OK,20000
   3,  60000,  6s,OK,20000
   3,  60000,  7s,OK,20000
   3,  60000,  7s,OK,20000
   3,  60000,  7s,OK,20000
   3,  60000,  7s,OK,20000
   3,  60000, 14s,OK,20000
   3,  60000, 18s,OK,20000
   3,  60000, 21s,OK,20000
   3,  60000, 23s,OK,20000
   3,  60000, 24s,OK,20000
   3,  60000, 32s,OK,20000
   3,  60000, 36s,OK,20000
   3,  60000, 37s,OK,20000
   3,  60000, 42s,OK,20000
   3,  60000, 42s,OK,20000
   3,  60000, 43s,OK,20000
   3,  60000, 43s,OK,20000
   3,  60000, 44s,OK,20000
   3,  60000, 44s,OK,20000
   3,  60000, 45s,OK,20000
   3,  60000, 46s,OK,20000
   3,  60000, 47s,OK,20000
   3,  60000, 47s,OK,20000
   3,  60000, 49s,OK,20000
   3,  60000, 49s,OK,20000
   3,  60000, 50s,OK,20000
   3,  60000, 50s,OK,20000
   3,  60000, 51s,OK,20000
   3,  60000, 53s,OK,20000
   3,  60000, 54s,OK,20000
   3,  60000, 55s,OK,20000
   3,  60000, 57s,OK,20000
   3,  60000, 58s,OK,20000
   3,  60000, 65s,OK,20000
   3,  60000, 66s,OK,20000
   3,  60000, 87s,OK,20000
   4,  80000, 10s,OK,20000
   4,  80000, 27s,OK,20000
   4,  80000, 40s,OK,20000
   4,  80000, 69s,OK,20000
   4,  80000, 71s,OK,20000
   5, 100000, 10s,OK,20000
   5, 100000, 16s,OK,20000
   5, 100000, 18s,OK,20000
   5, 100000, 75s,OK,20000
   6, 120000, 11s,OK,20000
   6, 120000, 82s,OK,20000
   6, 120000, 98s,OK,20000
   6, 120000, 99s,OK,20000
   7, 140000, 12s,OK,20000
   8, 160000,  8s,OK,20000
   9, 180000, 47s,OK,20000
   9, 180000,128s,OK,20000
  11, 220000, 36s,OK,20000
  11, 220000, 39s,OK,20000
  13, 260000, 22s,OK,20000
  17, 340000, 30s,OK,20000
  18, 360000, 36s,OK,20000
  19, 380000, 56s,OK,20000
  22, 440000, 41s,OK,20000
  22, 440000, 51s,OK,20000
  23, 460000, 83s,OK,20000
  26, 520000, 49s,OK,20000
  28, 560000,183s,OK,20000
  37, 740000,121s,OK,20000
  43, 860000,143s,OK,20000
  44, 880000,179s,OK,20000
  46, 920000, 81s,OK,20000
  46, 920000,162s,OK,20000
  50,1000000,114s,OK,20000
  53,1060000,131s,OK,20000
  54,1080000, 91s,OK,20000
  55,1100000,472s,FAIL,20000
  55,1100000,476s,FAIL,20000
  66,1320000,199s,OK,20000
  71,1420000,528s,FAIL,20000
  85,1700000,577s,FAIL,20000
  86,1720000,587s,FAIL,20000
  91,1820000,596s,FAIL,20000
  93,1860000,601s,FAIL,20000
 103,2060000,623s,FAIL,20000
 111,2220000,614s,FAIL,20000
 138,2760000,614s,FAIL,20000
 156,3120000,628s,FAIL,20000
 255,5100000,618s,FAIL,20000
 255,5100000,624s,FAIL,20000
  34,1325000,196s,OK,38970.6
  55,2170000,648s,FAIL,39454.5
 134,5325000,638s,FAIL,39738.8
   1,  40000,  8s,OK,40000
   2,  80000,  5s,OK,40000
   4, 160000, 12s,OK,40000
 101,4040000,627s,FAIL,40000
 101,4040000,633s,FAIL,40000
 109,4360000,654s,FAIL,40000
 171,6840000,634s,FAIL,40000
   1, 200000, 14s,OK,200000
   1, 200000, 14s,OK,200000
   1, 200000, 14s,OK,200000
[crab3@crab-prod-tw01 tmp]$ 

```

**Comment by yuyiguo** [comment]
@belforte 
The big lumi block insertion is case by case from the DB point of view. Because of the 300 second restriction imposed by FE, we have to get everything in DB with this limit. Current DBS python based server does multiple commits to the DB so re-insection of a failed larged block due to timeout will not start from the beginning. It may be inserted after the second try.  However, if the lumi number is too big for the DB to handle in any situation, this block will not be inserted or one has to reduce the size. 

My questions now for you are:
1. What is the minimum number of lumis/block or lumis/file for a certain type of data that requires a lot of lumis ?
2. Can crab accept async block insertion? That means that  DBS will email you the migration status after it is done.

**Comment by belforte** [comment]
Forgot to say: yes, I heard from you already that now we can try to insert a block multiple times and it will resume where it left. But current code does not retry failed blockinsert. What I saw in several cases is that HTTP call times out, so I get an error, but later on when I check I find that all files had been inserted in DBS, i.e. the server completed in the end, even if cmsweb FE timed out. This is anyhow rare enough that I did not go into putting automatic retry after block insert failure. 
I am not sure if I should and in case exactly how to do it, alwasy.. only after certain exceptions.. inserting a wait time.. a retry limit... all those annoying details.
Until now I thought it was all a matter of too-long-data-structures-which-will-eventually-be-solved and tried to wait it out. So many things to do already !

**Comment by yuyiguo** [comment]
I heard you. No rush here.

**Comment by belforte** [comment]
I tested the "1M lumis per block" limit on the pathological task and it seems to do the trick.
So am putting that in.

**Comment by belforte** [comment]
@yuyiguo I forgot to say that if user care, they can always issue `crab resubmit --publication` and that will force a retry of block insertion. The good thing about handing it to user to do, is that it guarantees a good delay from the time if fails, very convenient if DBS or cmsweb FE was having an unrelated issue. In a way CRAB has hundreds of (incompetent) operators, and I try to exploit the good side of that.
